### PR TITLE
Issue #56 - After successful login with token, use signed cookie to allow login with login step only

### PR DESCRIPTION
**Pull request by compsult**
(scoped to settings.LOGIN_URL). If cookie exists/is valid
on the next login, token steps are skipped

1 - if user checks checkbox to token step "Remember this device for %s days"
2 - create signed cookie with the desired end date 
3 - on next login, if cookie is present/valid, allow login with login step only

## Description
added a method token_required() that returns False when a valid 
signed cookie ('rememberdevice') is present. If not required, methods has_token_step()
and has_backup_step() will return None (skipping that logic).

To enable OTPRequiredMixin() to grant access, if token_required() 
returns False, a device will be added to user

## Motivation and Context

issue #56

## How Has This Been Tested?

yes, tests were added to test_views_login.test_with_backup_phone()
testing both the form change and the cookie logic.

Ubuntu 16.04, Postgres 9.3, Python 3.5.2
Django==2.0.3
django-bootstrap-form==3.4
django-debug-toolbar==1.9.1
django-formtools==2.1
django-otp==0.4.3
django-phonenumber-field==1.3.0
django-two-factor-auth==1.7.0

ran "make test" with no errors

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [X] My code follows the code style of this project. - I tried to :)
- [X] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly. (I will, if the changes are OK with you)
- [X] I have added tests to cover my changes.
- [X] All new and existing tests passed.

### CODE REVIEW

File: two_factor/views/core.py [hunk 1]
```diff
@@ -112,11 +125,23 @@ def done(self, form_list, **kwargs):
         if not is_safe_url(url=redirect_to, host=self.request.get_host()):
             redirect_to = resolve_url(settings.LOGIN_REDIRECT_URL)
 
+        response = redirect(redirect_to)
         device = getattr(self.get_user(), 'otp_device', None)
         if device:
             signals.user_verified.send(sender=__name__, request=self.request,
                                        user=self.get_user(), device=device)
-        return redirect(redirect_to)
+            if 'token-remember' in self.request.POST and \
```
Comment (Bouke) [inline two_factor/views/core.py]: This should not inspect the POST directly, but use the form data instead.

File: two_factor/views/core.py [hunk 2]
```diff
@@ -112,11 +125,23 @@ def done(self, form_list, **kwargs):
...
+                    self.request.POST['token-remember'] == "on" or \
+                    'backup-remember' in self.request.POST and \
+                    self.request.POST['backup-remember'] == "on":
+                login_good_until = str(date.today() +
+                    timedelta(days=settings.TWO_FACTOR_TRUSTED_DAYS))
+                response.set_signed_cookie(key='evl', value=login_good_until,
+                        salt=settings.TWO_FACTOR_SALT,
+                        max_age=settings.TWO_FACTOR_TRUSTED_DAYS * (3600 * 24),
+                    expires=login_good_until, path=settings.LOGIN_URL, domain=None,
```
Comment (Bouke) [inline two_factor/views/core.py]: Indentation is off here.

File: two_factor/views/core.py [hunk 3]
```diff
@@ -112,11 +125,23 @@ def done(self, form_list, **kwargs):
...
             signals.user_verified.send(sender=__name__, request=self.request,
                                        user=self.get_user(), device=device)
-        return redirect(redirect_to)
+            if 'token-remember' in self.request.POST and \
+                    self.request.POST['token-remember'] == "on" or \
+                    'backup-remember' in self.request.POST and \
+                    self.request.POST['backup-remember'] == "on":
+                login_good_until = str(date.today() +
+                    timedelta(days=settings.TWO_FACTOR_TRUSTED_DAYS))
+                response.set_signed_cookie(key='evl', value=login_good_until,
```
Comment (Bouke) [inline two_factor/views/core.py]: The cookie's name could more descriptive, e.g. `rememberdevice`.

File: two_factor/views/core.py [hunk 4]
```diff
@@ -71,11 +73,17 @@ class LoginView(IdempotentSessionWizardView):
     }
 
     def has_token_step(self):
-        return default_device(self.get_user())
+        if self.token_required(self.request):
```
Comment (Bouke) [inline two_factor/views/core.py]: This would read better if you make it a single statement: `return token_required(..) and default_device(..)`.

File: two_factor/views/core.py [hunk 5]
```diff
@@ -71,11 +73,17 @@ class LoginView(IdempotentSessionWizardView):
...
+            return default_device(self.get_user())
+        else:
+            return None
 
     def has_backup_step(self):
-        return default_device(self.get_user()) and \
-            'token' not in self.storage.validated_step_data
+        if self.token_required(self.request):
```
Comment (Bouke) [inline two_factor/views/core.py]: This would read better if you make it a single statement: `return token_required(..) and default_device(..) and 'token' not in ..`.

File: two_factor/forms.py [hunk 6]
```diff
@@ -141,6 +142,9 @@ class DisableForm(forms.Form):
 class AuthenticationTokenForm(OTPAuthenticationFormMixin, Form):
     otp_token = forms.IntegerField(label=_("Token"), min_value=1,
                                    max_value=int('9' * totp_digits()))
+    remember = forms.BooleanField(required=False,
+                                  label=_("Remember this device for %s days")
```
Comment (Bouke) [inline two_factor/forms.py]: This won't produce the expected translation for some languages. See also [pluralization](https://docs.djangoproject.com/en/2.0/topics/i18n/translation/#pluralization) in the Django documentation.

File: two_factor/views/core.py [hunk 7]
```diff
@@ -103,7 +107,12 @@ def done(self, form_list, **kwargs):
         """
         Login the user and redirect to the desired page.
         """
-        login(self.request, self.get_user())
+        # if no token is required, populate user.otp_device
+        # so OTPRequiredMixin will grant access
+        user = self.get_user()
+        if not self.token_required(self.request):
+            user.otp_device = self.get_device()
```
Comment (Bouke) [inline two_factor/views/core.py]: I'd rather not specify the user's default device here, but a fake placeholder device instead. This makes it possible to distinguish whether the user actually verified a second factor in this session.

**Comment by Bouke** [comment]
Also; as this package is build on django-otp, there's also [django-agent-trust](http://django-agent-trust-official.readthedocs.io/en/latest/) and the otp implementation [django-otp-agents](http://django-otp-agents.readthedocs.io/en/latest/) from the same author. Instead of re-implementing those packages, we could also investigate re-using those. What do you think?

File: two_factor/views/core.py [hunk 8]
```diff
@@ -200,6 +220,26 @@ def get_context_data(self, form, **kwargs):
             context['cancel_url'] = resolve_url(settings.LOGOUT_URL)
         return context
 
+    def token_required(self, request):
```
[the remark this answers is not in the document]
Comment (compsult) [reply two_factor/views/core.py]: Yes, the cookie is not linked to a user account. I'm looking into methods for that linkage

**Comment by diegobill** [comment]
Using this version I am receiving this error message sometimes after login (using the same username in diferents devices):

```
Traceback:

File "/usr/local/lib/python3.5/dist-packages/django/core/handlers/exception.py" in inner
  35.             response = get_response(request)

File "/usr/local/lib/python3.5/dist-packages/django/core/handlers/base.py" in _get_response
  128.                 response = self.process_exception_by_middleware(e, request)

File "/usr/local/lib/python3.5/dist-packages/django/core/handlers/base.py" in _get_response
  126.                 response = wrapped_callback(request, *callback_args, **callback_kwargs)

File "/usr/local/lib/python3.5/dist-packages/django/views/generic/base.py" in view
  69.             return self.dispatch(request, *args, **kwargs)

File "/usr/local/lib/python3.5/dist-packages/django/utils/decorators.py" in _wrapper
  62.             return bound_func(*args, **kwargs)

File "/usr/local/lib/python3.5/dist-packages/django/views/decorators/debug.py" in sensitive_post_parameters_wrapper
  76.             return view(request, *args, **kwargs)

File "/usr/local/lib/python3.5/dist-packages/django/utils/decorators.py" in bound_func
  58.                 return func.__get__(self, type(self))(*args2, **kwargs2)

File "/usr/local/lib/python3.5/dist-packages/django/utils/decorators.py" in _wrapper
  62.             return bound_func(*args, **kwargs)

File "/usr/local/lib/python3.5/dist-packages/django/views/decorators/cache.py" in _wrapped_view_func
  44.         response = view_func(request, *args, **kwargs)

File "/usr/local/lib/python3.5/dist-packages/django/utils/decorators.py" in bound_func
  58.                 return func.__get__(self, type(self))(*args2, **kwargs2)

File "/usr/local/lib/python3.5/dist-packages/django/utils/decorators.py" in _wrapper
  62.             return bound_func(*args, **kwargs)

File "/usr/local/lib/python3.5/dist-packages/django/views/decorators/debug.py" in sensitive_post_parameters_wrapper
  76.             return view(request, *args, **kwargs)

File "/usr/local/lib/python3.5/dist-packages/django/utils/decorators.py" in bound_func
  58.                 return func.__get__(self, type(self))(*args2, **kwargs2)

File "/usr/local/lib/python3.5/dist-packages/django/utils/decorators.py" in _wrapper
  62.             return bound_func(*args, **kwargs)

File "/usr/local/lib/python3.5/dist-packages/django/views/decorators/cache.py" in _wrapped_view_func
  44.         response = view_func(request, *args, **kwargs)

File "/usr/local/lib/python3.5/dist-packages/django/utils/decorators.py" in bound_func
  58.                 return func.__get__(self, type(self))(*args2, **kwargs2)

File "/usr/local/lib/python3.5/dist-packages/formtools/wizard/views.py" in dispatch
  248.         response = super(WizardView, self).dispatch(request, *args, **kwargs)

File "/usr/local/lib/python3.5/dist-packages/django/views/generic/base.py" in dispatch
  89.         return handler(request, *args, **kwargs)

File "/usr/local/lib/python3.5/dist-packages/two_factor/views/core.py" in post
  105.         return super(LoginView, self).post(*args, **kwargs)

File "/usr/local/lib/python3.5/dist-packages/two_factor/views/utils.py" in post
  145.         return super(IdempotentSessionWizardView, self).post(*args, **kwargs)

File "/usr/local/lib/python3.5/dist-packages/formtools/wizard/views.py" in post
  309.                 return self.render_done(form, **kwargs)

File "/usr/local/lib/python3.5/dist-packages/two_factor/views/utils.py" in render_done
  195.         done_response = self.done(final_form_list, **kwargs)

File "/u/django/ZZZ/XXX/YYY/views.py" in done
  687.         return super(LoginView, self).done(form_list, **kwargs)

File "/usr/local/lib/python3.5/dist-packages/two_factor/views/core.py" in done
  116.             user.otp_device.token_step_skipped = True

Exception Type: AttributeError at /YYY/login/
Exception Value: 'NoneType' object has no attribute 'token_step_skipped'
```

**Comment by compsult** [comment]
Thanks diegobill. I got derailed with something else and expect to return to this in about a week. Appreciate the stack trace. 

 

    On Tuesday, August 21, 2018, 6:09:55 AM EDT, diegobill <notifications@github.com> wrote:  
 
 
Using this version I am receiving this error message sometimes after login (using the same username in diferents devices):
Traceback:

File "/usr/local/lib/python3.5/dist-packages/django/core/handlers/exception.py" in inner
  35.             response = get_response(request)

File "/usr/local/lib/python3.5/dist-packages/django/core/handlers/base.py" in _get_response
  128.                 response = self.process_exception_by_middleware(e, request)

File "/usr/local/lib/python3.5/dist-packages/django/core/handlers/base.py" in _get_response
  126.                 response = wrapped_callback(request, *callback_args, **callback_kwargs)

File "/usr/local/lib/python3.5/dist-packages/django/views/generic/base.py" in view
  69.             return self.dispatch(request, *args, **kwargs)

File "/usr/local/lib/python3.5/dist-packages/django/utils/decorators.py" in _wrapper
  62.             return bound_func(*args, **kwargs)

File "/usr/local/lib/python3.5/dist-packages/django/views/decorators/debug.py" in sensitive_post_parameters_wrapper
  76.             return view(request, *args, **kwargs)

File "/usr/local/lib/python3.5/dist-packages/django/utils/decorators.py" in bound_func
  58.                 return func.__get__(self, type(self))(*args2, **kwargs2)

File "/usr/local/lib/python3.5/dist-packages/django/utils/decorators.py" in _wrapper
  62.             return bound_func(*args, **kwargs)

File "/usr/local/lib/python3.5/dist-packages/django/views/decorators/cache.py" in _wrapped_view_func
  44.         response = view_func(request, *args, **kwargs)

File "/usr/local/lib/python3.5/dist-packages/django/utils/decorators.py" in bound_func
  58.                 return func.__get__(self, type(self))(*args2, **kwargs2)

File "/usr/local/lib/python3.5/dist-packages/django/utils/decorators.py" in _wrapper
  62.             return bound_func(*args, **kwargs)

File "/usr/local/lib/python3.5/dist-packages/django/views/decorators/debug.py" in sensitive_post_parameters_wrapper
  76.             return view(request, *args, **kwargs)

File "/usr/local/lib/python3.5/dist-packages/django/utils/decorators.py" in bound_func
  58.                 return func.__get__(self, type(self))(*args2, **kwargs2)

File "/usr/local/lib/python3.5/dist-packages/django/utils/decorators.py" in _wrapper
  62.             return bound_func(*args, **kwargs)

File "/usr/local/lib/python3.5/dist-packages/django/views/decorators/cache.py" in _wrapped_view_func
  44.         response = view_func(request, *args, **kwargs)

File "/usr/local/lib/python3.5/dist-packages/django/utils/decorators.py" in bound_func
  58.                 return func.__get__(self, type(self))(*args2, **kwargs2)

File "/usr/local/lib/python3.5/dist-packages/formtools/wizard/views.py" in dispatch
  248.         response = super(WizardView, self).dispatch(request, *args, **kwargs)

File "/usr/local/lib/python3.5/dist-packages/django/views/generic/base.py" in dispatch
  89.         return handler(request, *args, **kwargs)

File "/usr/local/lib/python3.5/dist-packages/two_factor/views/core.py" in post
  105.         return super(LoginView, self).post(*args, **kwargs)

File "/usr/local/lib/python3.5/dist-packages/two_factor/views/utils.py" in post
  145.         return super(IdempotentSessionWizardView, self).post(*args, **kwargs)

File "/usr/local/lib/python3.5/dist-packages/formtools/wizard/views.py" in post
  309.                 return self.render_done(form, **kwargs)

File "/usr/local/lib/python3.5/dist-packages/two_factor/views/utils.py" in render_done
  195.         done_response = self.done(final_form_list, **kwargs)

File "/u/django/ZZZ/XXX/YYY/views.py" in done
  687.         return super(LoginView, self).done(form_list, **kwargs)

File "/usr/local/lib/python3.5/dist-packages/two_factor/views/core.py" in done
  116.             user.otp_device.token_step_skipped = True

Exception Type: AttributeError at /YYY/login/
Exception Value: 'NoneType' object has no attribute 'token_step_skipped'

File: two_factor/views/core.py [hunk 9]
```diff
@@ -103,7 +108,13 @@ def done(self, form_list, **kwargs):
         """
         Login the user and redirect to the desired page.
         """
-        login(self.request, self.get_user())
+        # if no token is required, populate user.otp_device
+        # so OTPRequiredMixin will grant access
+        user = self.get_user()
+        if not self.token_required(self.request):
+            user.otp_device = self.get_device()
+            user.otp_device.token_step_skipped = True
```
Comment (diegobill) [inline two_factor/views/core.py]: The variable token_step_skipped is declared but never used

**Comment by diegobill** [comment]
Hi @compsult ,

I found another bug, when the user enables the 2fa and set a device to be remember, after a user disable the 2fa and enable again the 2fa does not require the token, because the cookie rememberdevice keeps working.

**Comment by compsult** [comment]
> Hi @compsult ,
> 
> I found another bug, when the user enables the 2fa and set a device to be remember, after a user disable the 2fa and enable again the 2fa does not require the token, because the cookie rememberdevice keeps working.

Bouke,

    If I define Yubi column in the TrustedAgent table as (FK → otp_yubikey_remoteyubikeydevice), otp_yubikey becomes required instead of optional. 
However, this would have the advantage of cascading the delete to the TrustedAgent, so in the example diegobill cites, this would automatically disable all user agents associated with a given Yubi. 

If I just save the otp_yubikey_remoteyubikeydevice id field in the TrustedAgent table, the otp_yubikey app remains optional but it would up to admin to manually delete TrustedAgent's associated with the Yubi (which seems like an error prone process). 

Thoughts?

**Comment by compsult** [comment]
@diegobill I believe I have corrected the issues with the latest commit

**Comment by compsult** [comment]
@Bouke , I checked in the latest changes several hours ago but the coverage reports have not appeared. Did I do something wrong?

**Comment by moggers87** [comment]
I've restarted the travis build to see if that helps.

**Comment by compsult** [comment]
In using these changes, I've found that when 2 users share a machine, they overwrite the others' rememberdevice cookie, (forcing them to use a token again). 
One way to allow multiple users to have a trusted device (using the same device) could be to append the first letters of their username to the end of the cookie (like 'rememberdevice_joe'), then they would step on each others cookies. Just an idea

**Comment by compsult** [comment]
Explanation of Changes
----------------------

1. 'rememberdevice' cookie now requires back end validation to allow skip of token step (user and user_agent must match a row in the TrustedAgent table)

2. Yubi device is verified to still exist before allowing cookie to skip token step (this is not necessary for Phones, as TrustedAgent rows are automatically deleted when the Phone is deleted)

3. Optional email of 'New sign in to your account' when new device succeeds in logging in

4. disabling 2FA will clear out all TrustedDevices table rows for that account

5. stolen/lost devices can be rendered invalid by removing the row from PhoneDevice or RemoteYubiKey Devices

Now, if a bad actor gains a users credentials and is able to obtain the 'rememberdevice' cookie, s/he will only be able to use the cookie if s/he is using the same user agent associated with that cookie

**Comment by thebookie** [comment]
Are there any plans on adding this in the near future? Would be such a great addition to an already great package!

**Comment by lsemel** [comment]
This would be a great addition.

File: Makefile [hunk 10]
```diff
@@ -24,7 +24,7 @@ migrations:
 coverage:
 	coverage erase
 	DJANGO_SETTINGS_MODULE=tests.settings PYTHONPATH=. \
-		coverage run ---parallel --source=two_factor \
+		coverage run -p --source=two_factor \
```
Comment (moggers87) [inline Makefile]: Not sure why you've made this change.

File: example/forms.py [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+    class Meta:
+        model = User
+        fields = ("username", "email", "password1", "password2")
+
+    def save(self, commit=True):
+        user = super(UserCreateForm, self).save(commit=False)
+        user.email = self.cleaned_data["email"]
+        if commit:
+            user.save()
+        return user
```
Comment (moggers87) [inline example/forms.py]: I think we can get rid of this (see my comments on `TWO_FACTOR_NEW_DEV_ALERTS`)

File: two_factor/views/utils.py [hunk 12]
```diff
@@ -1,15 +1,46 @@
...
+
+
+def send_user_verified(request, user, device):
+    # prevent circular imports
+    from two_factor import signals
+    signals.user_verified.send(sender=None, request=request,
+                               user=user, device=device)
+
+
+def login_alerts(request):
```
Comment (moggers87) [inline two_factor/views/utils.py]: This really should not be part of this PR. It's nothing to do with skipping 2fa and assumes the user has an email address.

File: two_factor/models.py [hunk 13]
```diff
@@ -118,3 +118,25 @@ def generate_challenge(self):
...
+    Disabling 2FA will clear the TrustedAgent table for that user.
+    Stolen/lost phones can be handled by removing them from the PhoneDevice table.
+    Stolen/lost Yubi devices can be handled by removing them from the
+    Remote YubiKey device table
+    """
+    user = models.ForeignKey(getattr(settings, 'AUTH_USER_MODEL', 'auth.User'),
+                            null=False, on_delete=models.CASCADE)
+    user_agent = models.CharField(null=False, blank=True, max_length=200)
+    yubi_id = models.IntegerField(null=True)
+    phone = models.ForeignKey(PhoneDevice, on_delete=models.CASCADE, null=True)
```
Comment (moggers87) [inline two_factor/models.py]: What about other django-otp plugins? E.g. what happens if I lose my TOTP device?

You might want to use a GenericForeignKey here… but I thinik that would break delete cascades?

File: two_factor/models.py [hunk 14]
```diff
@@ -118,3 +118,25 @@ def generate_challenge(self):
...
+    ip = models.GenericIPAddressField(null=True, blank=True, verbose_name='IP')
+
+    class Meta:
+        unique_together = (('user', 'user_agent'))
```
Comment (moggers87) [inline two_factor/models.py]: What happens when I log in on both my desktop and laptop? They both have the same browser and OS.

File: two_factor/signals.py [hunk 15]
```diff
@@ -1,3 +1,12 @@
...
+from two_factor.views.core import LoginView
+from two_factor.views.utils import login_alerts
 
 user_verified = Signal(providing_args=['request', 'user', 'device'])
+login_alert = Signal(providing_args=['request'])
+
+
+@receiver(login_alert, sender=LoginView)
+def login_alert(sender, **kwargs):
+    login_alerts(kwargs['request'])
```
Comment (moggers87) [inline two_factor/signals.py]: See my comments on `TWO_FACTOR_NEW_DEV_ALERTS` below.

File: two_factor/templatetags/device_format.py [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+    (re.compile('NT 6.1'), _('Windows 7')),
+    (re.compile('NT 6.2'), _('Windows 8')),
+    (re.compile('NT 6.3'), _('Windows 8.1')),
+    (re.compile('NT 10.0'), _('Windows 10')),
+    (re.compile('Windows'), _('Windows')),
+)
+
+
+@register.filter
+def agent_format(value):
```
Comment (moggers87) [inline two_factor/templatetags/device_format.py]: Rather than maintain our own regexes, why don't we use something like https://github.com/ua-parser/uap-python ?

**Comment by moggers87** [comment]
Also, with regard to django-otp-agents: if there are features missing from that package which mean we can't use it, I think it would be better to contribute them there so we *can*.

**Comment by compsult** [comment]
On Friday, November 15, 2019, 02:52:59 PM EST, Matt Molyneaux <notifications@github.com> wrote:  
 
 
@moggers87 requested changes on this pull request.

There are quite a few changes that don't really belong in this PR! I agree

I've also made some comments on the TrustedAgent model

In Makefile:
> @@ -24,7 +24,7 @@ migrations:
 coverage:
 	coverage erase
 	DJANGO_SETTINGS_MODULE=tests.settings PYTHONPATH=. \
-		coverage run ---parallel --source=two_factor \
+		coverage run -p --source=two_factor \
Not sure why you've made this change.I can't remember why
In example/forms.py:
> +from django.utils.translation import ugettext_lazy as _
+
+
+class UserCreateForm(UserCreationForm):
+    email = forms.EmailField(label=_("Email address"), required=True, help_text=_("Required."))
+
+    class Meta:
+        model = User
+        fields = ("username", "email", "password1", "password2")
+
+    def save(self, commit=True):
+        user = super(UserCreateForm, self).save(commit=False)
+        user.email = self.cleaned_data["email"]
+        if commit:
+            user.save()
+        return user
I think we can get rid of this (see my comments on TWO_FACTOR_NEW_DEV_ALERTS)I added a form to create allow entry of email while registering

In two_factor/views/utils.py:
>  logger = logging.getLogger(__name__)
 
 
+def send_new_device(request):
+    # prevent circular imports
+    from two_factor import signals
+    signals.login_alert(sender=None, request=request)
+
+
+def send_user_verified(request, user, device):
+    # prevent circular imports
+    from two_factor import signals
+    signals.user_verified.send(sender=None, request=request,
+                               user=user, device=device)
+
+
+def login_alerts(request):
This really should not be part of this PR. It's nothing to do with skipping 2fa and assumes the user has an email address.The email is optional by setting TWO_FACTOR_NEW_DEV_ALERTS to False, so it does not assume an email address, though it’ hard to imagine allowing registration without one.


In two_factor/models.py:
> +class TrustedAgent(models.Model):
+    """
+    When a login comes in with a skip token cookie, a corresponding
+    TrustedAgent record will need to match both the user id and user_agent.
+    If there is no matching record in the TrustedAgent table,
+    the token step will be required, even if it is a valid cookie.
+    Disabling 2FA will clear the TrustedAgent table for that user.
+    Stolen/lost phones can be handled by removing them from the PhoneDevice table.
+    Stolen/lost Yubi devices can be handled by removing them from the
+    Remote YubiKey device table
+    """
+    user = models.ForeignKey(getattr(settings, 'AUTH_USER_MODEL', 'auth.User'),
+                            null=False, on_delete=models.CASCADE)
+    user_agent = models.CharField(null=False, blank=True, max_length=200)
+    yubi_id = models.IntegerField(null=True)
+    phone = models.ForeignKey(PhoneDevice, on_delete=models.CASCADE, null=True)
What about other django-otp plugins? E.g. what happens if I lose my TOTP device?I looked at other django-otp plugins (see comments above)

You might want to use a GenericForeignKey here… but I thinik that would break delete cascades?

In two_factor/models.py:
> +    If there is no matching record in the TrustedAgent table,
+    the token step will be required, even if it is a valid cookie.
+    Disabling 2FA will clear the TrustedAgent table for that user.
+    Stolen/lost phones can be handled by removing them from the PhoneDevice table.
+    Stolen/lost Yubi devices can be handled by removing them from the
+    Remote YubiKey device table
+    """
+    user = models.ForeignKey(getattr(settings, 'AUTH_USER_MODEL', 'auth.User'),
+                            null=False, on_delete=models.CASCADE)
+    user_agent = models.CharField(null=False, blank=True, max_length=200)
+    yubi_id = models.IntegerField(null=True)
+    phone = models.ForeignKey(PhoneDevice, on_delete=models.CASCADE, null=True)
+    ip = models.GenericIPAddressField(null=True, blank=True, verbose_name='IP')
+
+    class Meta:
+        unique_together = (('user', 'user_agent'))
What happens when I log in on both my desktop and laptop? They both have the same browser and OS.What you describe is an issue common to using a user agent subset. Better than nothing but not perfect


In two_factor/signals.py:
>  
 user_verified = Signal(providing_args=['request', 'user', 'device'])
+login_alert = Signal(providing_args=['request'])
+
+
+@receiver(login_alert, sender=LoginView)
+def login_alert(sender, **kwargs):
+    login_alerts(kwargs['request'])

See my comments on TWO_FACTOR_NEW_DEV_ALERTS below.

In two_factor/templatetags/device_format.py:
> +    (re.compile('Mac OS X 10[._]11'), _('OS X El Capitan')),
+    (re.compile('Mac OS X 10[._]12'), _('macOS Sierra')),
+    (re.compile('Mac OS X 10[._]13'), _('macOS High Sierra')),
+    (re.compile('Mac OS X'), _('OS X')),
+    (re.compile('NT 5.1'), _('Windows XP')),
+    (re.compile('NT 6.0'), _('Windows Vista')),
+    (re.compile('NT 6.1'), _('Windows 7')),
+    (re.compile('NT 6.2'), _('Windows 8')),
+    (re.compile('NT 6.3'), _('Windows 8.1')),
+    (re.compile('NT 10.0'), _('Windows 10')),
+    (re.compile('Windows'), _('Windows')),
+)
+
+
+@register.filter
+def agent_format(value):
Rather than maintain our own regexes, why don't we use something like https://github.com/ua-parser/uap-python ? 
TheRegex’s are an exact subset of the code from django-user-sessions



That said, I am going a different direction with 2FA but if I can offer any answers about this code, I'd be happy to

**Comment by moggers87** [comment]
> That said, I am going a different direction with 2FA but if I can offer any answers about this code, I'd be happy to

Does that mean you will no longer be working on this PR?

**Comment by compsult** [comment]
yes

 

    On Saturday, November 16, 2019, 06:39:46 AM EST, Matt Molyneaux <notifications@github.com> wrote:  
 
 

That said, I am going a different direction with 2FA but if I can offer any answers about this code, I'd be happy to


Does that mean you will no longer be working on this PR?

**Comment by moggers87** [comment]
In that case, I'll close this PR.
